pkg/installer/job.go

```diff
@@ -0,0 +1,263 @@
+package installer
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/redhat-appstudio/tssc/pkg/constants"
+	"github.com/redhat-appstudio/tssc/pkg/k8s"
+
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	applycorev1 "k8s.io/client-go/applyconfigurations/core/v1"
+	applymetav1 "k8s.io/client-go/applyconfigurations/meta/v1"
+	applyrbacv1 "k8s.io/client-go/applyconfigurations/rbac/v1"
+)
+
+// Job represents the assynchronous actor that runs a Job in the cluster to run
+// this installer container image on a pod. The idea is allowing a non-blocking
+// installation process for the MCP server.
+type Job struct {
+	kube    *k8s.Kube // kubernetes client
+	appName string    // common name for resources
+	retries int32     // job retries
+}
+
+// JobLabelSelector finds the unique installer job in the cluster.
+const JobLabelSelector = "installer.tssc-cli.redhat-appstudio.github.com"
+
+// JobState represents the state of the installer job in the cluster.
+type JobState int
+
+const (
+	// NotFound no installer job found in the cluster.
+	NotFound JobState = iota
+	// Deploying the installer job is running.
+	Deploying
+	// Failed the installer job has failed.
+	Failed
+	// Done the installer job has succeeded.
+	Done
+)
+
+// getJob retrieves the current state of the installer job. When not found it
+// returns a nil job.
+func (j *Job) getJob(ctx context.Context) (*batchv1.Job, error) {
+	bc, err := j.kube.BatchV1ClientSet("")
+	if err != nil {
+		return nil, err
+	}
+
+	jobList, err := bc.Jobs("").List(ctx, metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("type=%s", JobLabelSelector),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// When the job list is empty, it returns a nil job as well.
+	if len(jobList.Items) == 0 {
+		return nil, nil
+	}
+
+	// Returns error when multiple installer jobs are found in the cluster, with
+	// information to identify them.
+	if len(jobList.Items) > 1 {
```

Comment:
I don't see where the Job is cleaned after being run, does that mean that the Job can be run only once?

Reply:
Yes, for the time being. I'm going to work on the deploy job and send more PRs specifically about it.